source/beamline/BioXAS/BioXASCarbonFilterFarmActuatorControl.cpp

```diff
@@ -0,0 +1,333 @@
+{
+	return positionMap_.key(position, Window::None);
+}
+
+double BioXASCarbonFilterFarmActuatorControl::positionOfWindow(Window::Selection window) const
+{
+	return positionMap_.value(window, 0);
+}
+
+AMControl::FailureExplanation BioXASCarbonFilterFarmActuatorControl::move(double setpoint)
+{
+	// Check that the control is ready to move and the setpoint is valid.
+
+	if (!isConnected()) {
+		AMErrorMon::alert(this, BIOXAS_FILTER_FARM_ACTUATOR_NOT_CONNECTED, "Failed to move carbon filter farm actuator: the arm is not connected.");
+		return AMControl::NotConnectedFailure;
+	}
+
+	if (isMoving()) {
+		AMErrorMon::alert(this, BIOXAS_FILTER_FARM_ACTUATOR_ALREADY_MOVING, "Failed to move carbon filter farm actuator: the arm is already moving.");
+		return AMControl::AlreadyMovingFailure;
+	}
+
+	if (!validWindowSetpoint(setpoint)) {
+		AMErrorMon::alert(this, BIOXAS_FILTER_FARM_ACTUATOR_INVALID_SETPOINT, "Failed to move carbon filter farm actuator: invalid setpoint specified.");
+		return AMControl::LimitFailure;
+	}
+
+	// Update the setpoint.
+
+	setWindowSetpoint(setpoint);
+
+	// Create an action that moves the actuator to the desired window, and checks that the final position is okay.
+
+	AMAction3 *moveAction = createMoveAction(setpoint_);
+
+	if (!moveAction) {
+		AMErrorMon::alert(this, BIOXAS_FILTER_FARM_ACTUATOR_MOVE_FAILED, "Failed to move the carbon filter farm actuator.");
+		return AMControl::LimitFailure;
+	}
+
+	// Create move action signal mappings.
+
+	connect( moveAction, SIGNAL(started()), this, SLOT(onMoveStarted()) );
+
+	cancelledMapper_->setMapping(moveAction, moveAction);
+	connect( moveAction, SIGNAL(cancelled()), cancelledMapper_, SLOT(map()) );
+
+	failedMapper_->setMapping(moveAction, moveAction);
+	connect( moveAction, SIGNAL(failed()), failedMapper_, SLOT(map()) );
+
+	succeededMapper_->setMapping(moveAction, moveAction);
+	connect( moveAction, SIGNAL(succeeded()), succeededMapper_, SLOT(map()) );
+
+	// Run action.
+
+	moveAction->start();
+
+	return AMControl::NoFailure;
+}
+
+void BioXASCarbonFilterFarmActuatorControl::setWindowPosition(Window::Selection window, double position)
+{
+	positionMap_.insert(window, position);
+}
+
+void BioXASCarbonFilterFarmActuatorControl::setWindow(double newWindow)
+{
+	if (value_ != newWindow) {
+		value_ = newWindow;
+		emit valueChanged(value_);
+	}
+}
+
+void BioXASCarbonFilterFarmActuatorControl::setWindowSetpoint(double newWindow)
+{
+	if (setpoint_ != newWindow) {
+		setpoint_ = newWindow;
+		emit setpointChanged(setpoint_);
+	}
+}
+
+void BioXASCarbonFilterFarmActuatorControl::setMoveInProgress(bool isMoving)
+{
+	if (moveInProgress_ != isMoving) {
+		moveInProgress_ = isMoving;
+		emit movingChanged(moveInProgress_);
+
+		if (moveInProgress_)
+			emit moveStarted();
+	}
+}
+
+void BioXASCarbonFilterFarmActuatorControl::onMoveStarted()
+{
+	setMoveInProgress(true);
```

Comment:
It should have been called in the moveCleanup() methods of both the actuator control and the general filter farm control, but was missing. Should be included now.